docs/pages/pmd/devdocs/testing.md

```diff
@@ -1,6 +1,263 @@
+the test code on its own.
+
+The test code (see below [Test XML Reference](#test-xml-reference)) describes the test case completely with
+the expected behavior like number of expected rule violations, where the violations are expected, and so on.
+
+When you are running the test class in your IDE (e.g. Eclipse or IntelliJ IDEA) you can also select a single
+test case and just execute this one.
+
+## Where to place the test code
+
+The `SimpleAggregatorTst` class searches the XML file, that describes the test cases for a certain rule
+using the following convention:
+The XML file is a test resource, so it is searched in the tree under `src/test/resources`.
+
+The sub package `xml` of the test class's package should contain a file with the same name as the rule's name
+which is under test.
+
+For example, to test the "Java Basic Ruleset", the fully qualified test class is:
+
+    net.sourceforge.pmd.lang.java.rule.basic.BasicRulesTest
+
+The test code for the rule "AvoidBranchingStatementAsLastInLoop" can be found in the file:
+
+    src/test/resources/net/sourceforge/pmd/lang/java/rule/basic/xml/AvoidBranchingStatementAsLastInLoop.xml
+
+In general, the class name and file name pattern for the test class and data is this:
+
+    net.sourceforge.pmd.lang.<Language Terse Name>.rule.<Ruleset Name>.<Ruleset Name>RulesTest
+    src/test/resources/net/sourceforge/pmd/lang/<Language Terse Name>/rule/<Ruleset Name>/xml/<Rule Name>.xml
+
+{%include tip.html content="This convention allows you to quickly find the test cases for a given rule:
+Just search in the project for a file `<RuleName>.xml`. Looking at the path of the file, you can figure
+out the ruleset name. Searching for a class `<Ruleset Name>RulesTest` gives you the test class." %}
+
+## Simple example
+
+### Test Class: BasicRulesTest
+
+This is a stripped down example for the Java Basic Ruleset:
+
+``` java
+package net.sourceforge.pmd.lang.java.rule.basic;
+
+import net.sourceforge.pmd.testframework.SimpleAggregatorTst;
+
+public class BasicRulesTest extends SimpleAggregatorTst {
+
+    private static final String RULESET = "java-basic";
+
+    @Override
+    public void setUp() {
+        addRule(RULESET, "AvoidBranchingStatementAsLastInLoop");
+        addRule(RULESET, "AvoidDecimalLiteralsInBigDecimalConstructor");
+    }
+}
+```
+
+This test class overrides the method `setUp` in order to register test cases for the two rules. If there
+are more rules, just add additional `addRule(...)` calls.
+
+{%include note.html content="The `RULESET` constant points to `java-basic` which is just a shortcut
+for `rulesets/java/basic.xml`. " %}
+
+{%include note.html content="You can also add additionally standard JUnit test methods annotated with `@Test` to
+this test class." %}
+
+### Test Data: AvoidBranchingStatementAsLastInLoop.xml
+
+This is a stripped down example which just contains two test cases.
+
+``` xml
+<?xml version="1.0" encoding="UTF-8"?>
```

Comment:
we should provide an XML schema for test files. This would simplify documentation, allow us to validate files, and would allow IDEs to provide autocomplete.

Reply:
+1

Reply:
@jsotuyod Just wanted to let you know, I'm working on a schema. So far it looks like this: http://pmd.sourceforge.net/rule-tests_1_0_0.xsd

Reply:
@adangel sweet! Looking forward to it!